stream_alert/alert_processor/outputs/credentials/provider.py

```diff
@@ -0,0 +1,589 @@
+                enc_creds = file_handle.read()
+
+            credentials = Credentials(enc_creds, True, self._region)
+            if self._cache_driver:
+                self._cache_driver.save_credentials(descriptor, credentials)
+
+            return credentials
+        except ClientError as err:
+            LOGGER.exception('credentials for \'%s\' could not be downloaded '
+                             'from S3: %s',
+                             get_formatted_output_credentials_name(self._service_name, descriptor),
+                             err.response)
+
+    def has_credentials(self, descriptor):
+        """Always returns True, as S3 is the place where all encrypted credentials are
+           guaranteed to be cold-stored."""
+        return True
+
+    def save_credentials_into_s3(self, descriptor, credentials, kms_key_alias):
+        """
+        Takes the given credentials, encrypts them, and saves them into a determined S3 bucket
+        and location.
+
+        Notably, this implementation is NOT for the CredentialsCachingDriver interface, as the
+        S3Driver is not a caching driver.
+
+        Args:
+            descriptor (str):
+            credentials (Credentials):
+            kms_key_alias (str):
+
+        Returns:
+            bool: True on success, False otherwise.
+        """
+        if credentials.is_encrypted():
+            # Don't try saving encrypted credentials, or you will doubly-encrypt them, since this
+            # method will encrypt the credentials immediately prior to saving into S3.
+            raise RuntimeError('Cannot save encrypted credentials')
+
+        s3_key = get_formatted_output_credentials_name(self._service_name, descriptor)
+
+        # Encrypt the creds and push them to S3
+        # then update the local output configuration with properties
+        creds = credentials.data()
+        if not creds:
+            return True
+
+        creds_json = json.dumps(creds)
+        enc_creds = kms_encrypt(self._region, creds_json, kms_key_alias)
+        return send_creds_to_s3(self._region, self._bucket, s3_key, enc_creds)
+
+    def get_s3_key(self, descriptor):
+        """
+        Returns an appropriate S3 bucket key for credentials relevant to this Output
+
+        Args:
+            descriptor (str):
+
+        Returns:
+            string
+        """
+        return get_formatted_output_credentials_name(self._service_name, descriptor)
+
+    def get_s3_secrets_bucket(self):
+        """
+        Returns an appropriate S3 bucket for all credentials relevant to this driver.
+
+        Returns:
+            string
+        """
+        return '{}.streamalert.secrets'.format(self._prefix)
+
+
+class LocalFileDriver(CredentialsProvidingDriver, FileDescriptorProvider, CredentialsCachingDriver):
+    """Driver for fetching credentials that are saved locally on the filesystem."""
+
+    def __init__(self, region, service_name):
+        self._region = region
```

Comment:
I don't think region will ever be used for the 'local' one right?

Reply:
Unfortunately I need it to pass into the `Credentials` constructor, because that class requires the `region` to use KMS to decrypt the credentials. I honestly could've passed in a `CredentialsFactory` kind of thing that would avoid having to hot-potato around all of these `region`, `service_name` and so on things.

Reply:
but only the `S3` driver would need to do the decryption right? or am I misunderstanding this?

if that's the case - the S3 driver constructor can take different params than the base (super) class and just call the super's init with the right params also. I could be misunderstanding the inheritance model, though

Reply:
Not exactly; the S3 driver (none of the "drivers") actually do any decryption on their own. I delegated **decryption** to this `Credentials` class.

I didn't actually use any traditional inheritance. The three "base" classes are:

* [FileDescriptorProvider](https://github.com/airbnb/streamalert/pull/878/files#diff-4463a710593722ea77bc4209c8fc8fcdR202)
* [CredentialsCachingDriver](https://github.com/airbnb/streamalert/pull/878/files#diff-4463a710593722ea77bc4209c8fc8fcdR216)
* [CredentialsProvidingDriver](https://github.com/airbnb/streamalert/pull/878/files#diff-4463a710593722ea77bc4209c8fc8fcdR172) (The important one)

The implementations of `CredentialsProvidingDriver` are supposed to return `Credentials`, which can be `is_encrypted()` or not. The idea is (was?) to allow credentials to be passed between drivers while still encrypted. The only thing the driver needs to know is whether or not `Credentials` are encrypted "at rest" or not.

This is why all of the Drivers need a `region`; because they need to pass it to the `Credentials` class to be able to decrypt the ciphertext after being read from disk/memory/whatever. The exception is the `EphemeralDriver`, which is the only unencrypted-at-rest Driver.